river/cluster/denstream.py

```diff
@@ -0,0 +1,491 @@
+                    # delete c_o
+                    self.o_micro_clusters.pop(j)
+        return self
+
+    def predict_one(self, x, sample_weight=None):
+        """Predict the cluster index for each sample
+
+        `predict_one()` is used to handle the case when a clustering request arrives.
+        When a clustering request arrives, a variant of DBSCAN algorithm [^2] is applied on the set of on-line
+        maintained p-micro-clusters to get the final result of clustering. Each p-micro-cluster `c_p`
+        is regarded as a virtual point localed at the center of `c_p` with weight $\\omega$.
+        After new clusters of p-micro-clusters are generated, the label of the new point is obtained
+        by assessing the nearest cluster center with respect to that specific point using the already implemented
+        query function `_find_closest_cluster_index()`.
+
+        Parameters
+        ----------
+        x
+            The new point instance
+
+        sample_weight
+            Instance weight
+
+        Returns
+        -------
+        y
+            Cluster label
+
+        """
+        # This function handles the case when a clustering request arrives.
+        # implementation of the DBSCAN algorithm proposed by Ester et al.
+
+        # initiate labels of p-micro-clusters to None
+        labels = {pmc: None for pmc in self.p_micro_clusters.values()}
+
+        # cluster counter; in this algorithm cluster labels start with 0
+        c = -1
+
+        for pmc in self.p_micro_clusters.values():
+            # previously processed in inner loop
+            if labels[pmc] is not None:
+                continue
+            pmc_neighbors = self._query_neighbor(pmc)
+            # no label as noise, as there are no min points
+            # next cluster label
+            c += 1
+            labels[pmc] = c
+            # neighbors to expand has already been generated by query_neighbor
+            # which means that pmc_neighbors is already a seed set
+            seed_set = pmc_neighbors.copy()
+            # process every point in seed set
+            for neighbor in seed_set:
+                # check previously proceeded points
+                if labels[neighbor] is not None:
+                    continue
+                labels[neighbor] = c
+                # find neighbors
+                neighbor_neighbors = self._query_neighbor(neighbor)
+                # add new neighbors to seed set
+                # no need for density check, as there are no minPts
+                for neighbor_neighbor in neighbor_neighbors.keys():
+                    seed_set[neighbor_neighbor] = None
+
+        self.n_clusters, self.clusters = self._generate_clusters_from_labels(labels)
+
+        self.centers = {i: self.clusters[i].center for i in range(self.n_clusters)}
+
+        y = self._find_closest_cluster_index(x, self.clusters)
+
+        return y
+
+
+class MicroCluster(metaclass=ABCMeta):
```

Comment:
Since we are exposing this class then it is worth differentiating it from the CluStream micro-clusters. What about `DSMicroCluster`, `DenStreamMicroCluster`?

Reply:
I would call it `DenStreamMicroCluster` to avoid any shadowing that may arise (as we also have DStream that we can consider implementing it later).

Reply:
Nice